packages/playground-examples/copy/pt/JavaScript/Functions with JavaScript/Generic Functions.ts

```diff
@@ -0,0 +1,100 @@
+
+// Vamos começar com um exemplo onde envolveremos um objeto de entrada
+// em um array. Apenas nos importaremos com uma só variável neste
+// caso: o tipo que foi passado como argumento:
+
+function envolverEmArray<Tipo>(entrada: Tipo): Tipo[] {
+  return [entrada];
+}
+
+// Nota: é comum ver Tipo ser referido como T. Isso é
+// culturalmente similar à como as pessoas usam i em um loop for
+// para representar índice. T normalmente representa Tipo, então
+// usaremos o nome completo para maior clareza
+
+// Nossa função usará inferência para para sempre manter o tipo
+// passado como argumento igual ao tipo retornado (porém
+// envolvido em um array)
+
+const stringArray = envolverEmArray("hello generics");
+const numberArray = envolverEmArray(123);
+
+// Podemos verificar que isso funciona como esperado ao checar
+// se podemos atribuir um array de strings à uma função que
+// deveria ser um array de objetos
+const naoArrayDeStrings: string[] = envolverEmArray({});
+
+// Você também pode evitar a inferência adicionando
+// o tipo você mesmo:
+const arrayDeStrings2 = envolverEmArray<string>("");
+
+// envolverEmArray permite que qualquer tipo seja usado, porém existem
+// casos onde você precisa permitir apenas um subconjunto de tipos.
+// Nesses casos você pode dizer que o tipo deve estender um
+// tipo específico.
+
+interface Desenhavel {
+  desenhar: () => void;
+}
+
+// Esta função receve um conjunto de objetos os quais possuem uma função
+// para desenhar na tela
+function renderizarNaTela<Tipo extends Desenhavel>(entrada: Tipo[]) {
+  entrada.forEach(i => i.desenhar());
+}
+
+const objetosComDesenhar = [{ desenhar: () => {} }, { desenhar: () => {} }];
+renderizarNaTela(objetosComDesenhar);
+
+// Isso falhará se desenhar não estiver present:
+
+renderizarNaTela([{}, { desenhar: () => {} }]);
+
+// Tipos genéricos podem começar a parecer complicados quando você tem
+// múltiplas variáveis. Aqui está um exemplo de uma função de caching
+// que permite que você tenha diferentes conjuntos de tipos de entrada
+// e caches.
+
+interface HostDeCache {
+  salvar: (a: any) => void;
+}
+
+function adicionarObjetoAoCache<Tipo, Cache extends HostDeCache>(obj: Tipo, cache: Cache): Cache {
+  cache.salvar(obj);
+  return cache;
+}
+
+// Este é o mesmo exemplo que acima, porém com um parâmetro extra.
+// Nota: para fazê-lo funcionar, porém, tivemos que usar any. Isso
+// pode ser resolvido usando uma interface genérica
+
+interface HostDeCacheGenerico<TipoDeConteudo> {
+  salvar: (a: TipoDeConteudo) => void;
+}
+
+// Agora quando o HostDeCacheGenerico é usado, você deve dizer
+// à dele qual é o TipoDeConteudo
+
+function adicionarObjetoTipadoAoCache<Tipo, Cache extends HostDeCacheGenerico<Tipo>>(obj: Tipo, cache: Cache): Cache {
+  cache.salvar(obj);
+  return cache;
+}
+
+// O exemplo acima é bem intenso em termos de sintaxe. Porém,
+// isso provê uma segurança maior. Estes são trade-offs (comprometimentos),
```

Comment:
```suggestion
// isso provê uma segurança maior. Estes são trade-offs (compromissos),
```
https://www.linguee.com/english-portuguese/search?source=auto&query=trade-off